File: Facebook/111_compressStr.py
# https://leetcode.com/problems/string-compression/

class Solution:
    def compress(self, chars: List[str]) -> int:
        if len(chars) == 0 : return chars
#get len
        llens=[]
        cc=[]
        prev=chars[0]
        ll = 1
        for i in range(1, len(chars)) :
            curr = chars[i]
            if curr == prev :
                ll += 1
            else :
                llens.append(ll)
                cc.append(prev)
                prev=curr
                ll=1

        cc.append(prev)
        llens.append(ll)        

#compress
        chars.clear()
        for i in range(0, len(cc)) :
            chars.append(cc[i])

            ll = llens[i]
            if ll > 1 :
                for x in list("%d" % ll) :
                    chars.append(x)

        return len(chars)

File: Facebook/111_CompressStr.java
// https://leetcode.com/problems/string-compression/

class Solution {
    public int compress(char[] str)
    {
        if (str.length < 2) return str.length;
        
        int trunc = 0, ll = 1;
        char prev = str[trunc];
        for (int j = 1; j < str.length; j++)
        {
            char curr = str[j];
            
            if (curr != prev)
            {
                trunc = helper(str, trunc, prev, ll);
                prev = curr;
                ll = 1;
            }
            else ll++;
        }

        return helper(str, trunc, prev, ll);
    }    

    private int helper(char[] string, int start, char key, int howmany)
    {
        string[start++] = key;
        if (howmany > 1)
        {
            for (char dd : Integer.toString(howmany).toCharArray())
                string[start++] = dd;
        }
        
        return start;
    }
}